src/include/api/ivf_flat_index.h

```diff
@@ -0,0 +1,707 @@
+      }
+      uint32_t count;
+      void* addr;
+      read_group.get_metadata(name, &datatype, &count, (const void**)&addr);
+      if (datatype == TILEDB_UINT32) {
+        *reinterpret_cast<uint32_t*>(value) =
+            *reinterpret_cast<uint32_t*>(addr);
+      } else {
+        throw std::runtime_error("Unsupported datatype for metadata: " + name);
+      }
+    }
+
+    /**
+     * We support all combinations of the following types for feature,
+     * id, and px datatypes:
+     *   feature_type: uint8 or float
+     *   id_type: uint32 or uint64
+     *   px_type: uint32 or uint64
+     *
+     *   @todo Unify the type-based switch-case statements in a manner
+     *   similar to what was done in query_condition
+     */
+    if (feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<uint8_t, uint32_t, uint32_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<float, uint32_t, uint32_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<uint8_t, uint32_t, uint64_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<float, uint32_t, uint64_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<uint8_t, uint64_t, uint32_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<float, uint64_t, uint32_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<uint8_t, uint64_t, uint64_t>>>(
+          ctx, group_uri, config);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<ivf_flat_index<float, uint64_t, uint64_t>>>(
+          ctx, group_uri, config);
+    }
```

Comment:
(Optional) Might be nice to add an else case and throw an exception. That way if we add to the enum and not here we can catch it more easily. Same with the other big if/else if below.